### PR TITLE
DDFFORM 313

### PR DESCRIPTION
**Pull request by LasseStaus**
#### Link to issue

[DDFFORM-313](https://reload.atlassian.net/browse/DDFFORM-313)

- CMS PR: https://github.com/danskernesdigitalebibliotek/dpl-cms/pull/796
- React PR: https://github.com/danskernesdigitalebibliotek/dpl-react/pull/971

#### Description

This PR add the design for Material Grid, showing a number of recommendedMaterials in a grid.  

#### Screenshot of the result

image

#### Additional comments or questions

I am not fond of the placement of the title, compared to the width of the component. This should be the correct width's however.

### CODE REVIEW

File: src/stories/Library/material-grid/MaterialGrid.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  materials,
+}) => {
+  useEffect(() => {
+    require("./init-material-grid");
+  }, []);
+
+  const showMoreButton = materials.length > 8;
+
+  return (
+    <div className={clsx("material-grid")}>
```
Comment (kasperbirch1) [inline src/stories/Library/material-grid/MaterialGrid.tsx:26]: Should this be a section for semantic reasons? and then I see no reason to use clsc when there is only one class?

File: src/stories/Library/material-grid/MaterialGrid.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      <div className="material-grid__title">{title}</div>
```
Comment (kasperbirch1) [inline src/stories/Library/material-grid/MaterialGrid.tsx:27]: I also think this should be a h tag

File: src/stories/Library/material-grid/MaterialGrid.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      <div className="material-grid__items">
```
Comment (kasperbirch1) [inline src/stories/Library/material-grid/MaterialGrid.tsx:28]: ul

File: src/stories/Library/material-grid/MaterialGrid.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        {materials.slice(0, amountOfMaterialsToShow).map((material, index) => (
+          <div
```
Comment (kasperbirch1) [inline src/stories/Library/material-grid/MaterialGrid.tsx:30]: li

File: src/stories/Library/material-grid/init-material-grid.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+window.addEventListener("load", () => {
```
Comment (kasperbirch1) [inline src/stories/Library/material-grid/init-material-grid.js:1]: I'm not sure if it's necessary but maybe we should use a DOMContentLoaded here

File: src/stories/Library/material-grid/init-material-grid.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  const materialGrids = document.querySelectorAll(
+    ".material-grid:not(.initialized)"
+  );
+
+  materialGrids.forEach((grid) => {
+    grid.classList.add("initialized");
+    const showMoreButton = grid.querySelector(".material-grid__show-more");
```
Comment (kasperbirch1) [inline src/stories/Library/material-grid/init-material-grid.js:8]: Is it better to capture elements with data attributes instead of class?

**Review by kasperbirch1 — APPROVED** [review]
I have some suggestions you could consider but or it looks fine to me 👍 

Remember to accept the UI changes in Chromatic before merging.

File: src/stories/Library/material-grid/MaterialGrid.tsx (hunk 2 above)
Comment (LasseStaus) [reply src/stories/Library/material-grid/MaterialGrid.tsx:27]: Used a <h2>

File: src/stories/Library/material-grid/init-material-grid.js (hunk 5 above)
Comment (LasseStaus) [reply src/stories/Library/material-grid/init-material-grid.js:1]: I'm not sure if everything is loaded at that time once used in the drupal setup, but we can test it with DOMContent first and change it later if it's a problem. 👍

File: src/stories/Library/material-grid/MaterialGrid.tsx (hunk 1 above)
Comment (LasseStaus) [reply src/stories/Library/material-grid/MaterialGrid.tsx]: We add a section around it with the paragraph rendering in drupal. But yes 👍

File: src/stories/Library/material-grid/MaterialGrid.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    },
+    href: {
+      control: "text",
+      description: "Link to the recommended material",
+    },
+    positionImageRight: {
+      control: "boolean",
+      description: "Toggle image position",
+    },
+  },
```
Comment (kasperg) [inline src/stories/Library/material-grid/MaterialGrid.stories.tsx:25]: Please reconsider these arguments.

Most of them are not relevant for this component. The only one that is - `title` has a somewhat misleading description.

File: src/stories/Library/material-grid/MaterialGrid.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+export const MaterialGrid: React.FC<MaterialGridProps> = ({
+  title,
+  amountOfMaterialsToShow,
+  materials,
+}) => {
+  // The JS code here is just an example of how this works.
+  // The proper implementation is in the react repo.
+  const [showAllMaterials, setShowAllMaterials] = useState(false);
```
Comment (kasperg) [inline src/stories/Library/material-grid/MaterialGrid.tsx:20]: :+1: Nice commenting.

File: src/stories/Library/material-grid/MaterialGrid.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  const [displayedItemCount, setDisplayedItemCount] = useState(8);
+  const showMoreButton = materials.length > 8;
```
Comment (kasperg) [inline src/stories/Library/material-grid/MaterialGrid.tsx:22]: Consider your use of the value 8.

- Here the number 8 is used specifically. The ticket says that the default is 4. Figma shows 8 (but without a "Show more" button. I wonder what is right.

File: src/stories/Library/material-grid/MaterialGrid.tsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+          <li key={index} className="material-grid__item">
+            <RecommendedMaterial {...material} />
+          </li>
+        ))}
+      </ul>
+      {showMoreButton && !showAllMaterials && (
+        <button
+          className="material-grid__show-more btn-primary btn-outline btn-medium"
+          data-show-more
+          aria-expanded="false"
```
Comment (kasperg) [inline src/stories/Library/material-grid/MaterialGrid.tsx:43]: Please consider add an `aria-controls` attribute to the list.

:+1: remembering to add this `aria-expanded` attribute but [according to MDN you should also add an `aria-controls` attribute with the id of the element which is expanded when clicking the button](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded#buttons).

File: src/stories/Library/material-grid/MaterialGridData.tsx [hunk 11]
```diff
@@ -0,0 +1,198 @@
+import { RecommendedMaterialProps } from "../recommended-material/RecommendedMaterial";
+
+const MaterialGridData: RecommendedMaterialProps[] = [
+  {
+    author: "Daniel Lee (2023)",
+    description: "Advanced Woodworking",
+    favoriteFill: false,
+    coverUrl: "#",
+    src: "images/book_cover_6.jpg",
+    alt: "book cover",
```
Comment (kasperg) [inline src/stories/Library/material-grid/MaterialGridData.tsx:10]: Consider using an empty alt text. 

I see the cover images as descriptive. They provide no additional information compared to what is already available in the DOM. In these cases providing an empty `alt` text is the right thing to do.

File: src/stories/Library/material-grid/material-grid.scss [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+}
+.material-grid__items {
+  display: grid;
+  grid-template-columns: repeat(2, 1fr);
+}
+
+.material-grid__items .recommended-material {
+  max-width: unset;
+  border: 1px solid $color__global-tertiary-1;
+}
```
Comment (kasperg) [inline src/stories/Library/material-grid/material-grid.scss:20]: Please look into how we can avoid referencing another component.

According to our BEM guidelines one block should not modify the style of another block. Either we should look into if we can add modifiers to `.recommended-material` to make it behave like we want or we should create a `.material-grid__item` element within this block, style it appropriately and use this class on the `.recommended-material` elements.

**Review by kasperg — CHANGES_REQUESTED** [review]
:+1: Nice but I think a few things needs a second look.

File: src/stories/Library/material-grid/MaterialGrid.stories.tsx (hunk 7 above)
Comment (LasseStaus) [reply src/stories/Library/material-grid/MaterialGrid.stories.tsx:25]: Yea this is definitely incorrect, i shouldn't have missed this. 🙃

File: src/stories/Library/material-grid/MaterialGrid.tsx (hunk 9 above)
Comment (LasseStaus) [reply src/stories/Library/material-grid/MaterialGrid.tsx:22]: I understand. I've refactored the code and the naming of variables. 

From my understanding the initial max before showing a button should be 8, but i see they mentioned 4 in the ticket now. 

Change it to be using 4 now, and updated the figma link to the the specific component that fits description.

**Review by kasperg — APPROVED** [review]
:+1: LGTM

**Pull request merged**